test/integration/http_integration.cc

```diff
@@ -1428,10 +1429,6 @@ void HttpIntegrationTest::testUpstreamDisconnectWithTwoRequests() {
   auto response = codec_client_->makeRequestWithBody(default_request_headers_, 1024);
   waitForNextUpstreamRequest();
 
-  // Request 2.
```

Comment:
> HttpIntegrationTest::testUpstreamDisconnectWithTwoRequests was racy. It was possible before for
both requests to be load balanced and sent to the first upstream connection. When the 200
for the first request is received, we would then disconnect the upstream. This would then 503
back. Not sure if we've lost some coverage by these changes; the alternative might be to configure
retries?

Hmm yeah, TBH I don't remember the original intention of this test. If I had to guess it's to make sure that the disconnect flow works correctly when we have 2 in flight connections, so this is changing things a bit. Don't we only have a single upstream and a single worker? What's the failing interleaving here?

Reply:
I think this is the sequence. Consider this for HTTP/2, where we have a single connection:

1. Client sends request (1) and (2) to Envoy.
2. Envoy sends requests (1) and (2) to upstream.
3. Upstream waits for a request, 200s request (1) and then disconnects. Request (2) might be queued up in the upstream.
4. Envoy sends back to client the 200 for request (1). However, it sees the disconnect and 502s request (2).

This breaks the test. I think the usual story is that the upstream disconnects in step 3 before Envoy has managed to send request (2) to the upstream. While we only have one worker an one upstream thread, these are both independent. New streams are autonomously accepted by the `FakeUpstream`, without any interlock with the test/client thread.

Reply:
OK yeah that makes sense for HTTP/2. Is this only flaking for HTTP/2? 

TBH I really don't recall all the history here. @alyssawilk do you? In a perfect world, I think we would actually explicitly test both scenarios (both the 200 happy case and the 503 unhappy case), but that might require splitting on HTTP/1 and HTTP/2 and probably using some stats waiting to correctly synchronize. @alyssawilk thoughts on what to do here?

Reply:
Yeah,I think there is still a window in (2) for this to happen with keep-alive and HTTP/1.1. Upstream 200s, Envoy asynchronously forwards the next request to the upstream and it's in-flight, then the disconnect happens.

I think we should resolve this one quickly, this is behind a lot of the test failures we're seeing on CI AFAICT.

Reply:
I think you can fix for HTTP/1.1 pretty easily by forcing Envoy to make 2 upstream connections. I think you can just wait for the 2 upstream connections before proceeding.

Reply:
I'm not sure of the original intention of the test. @lizan?

Reply:
I just spent a little more time looking at the original change and the test and I think I understand the intention. If I understand correctly, I think the test could be done as follows:
1) Do test only for HTTP/1.1
2) Set max upstream connections to 1
3) Issue 2 requests on 2 connections
4) Use stats to make sure 1 request is pending and 1 is active
5) Respond and disconnect
6) Make sure 2nd request goes through.

If this is too complicated to fix quickly and you want to deflake, it's probably fine to comment out this test and fix in a follow up? Or maybe @lizan can fix?

Reply:
+1 - I'm fine with DISABLING_ and adding a TODO for later

Reply:
I will fix this today, I'm in an all-day meeting so it might be a bit delayed.